src/community/ardrive_contract_oracle.ts

```diff
@@ -0,0 +1,155 @@
+const initialContractReader = 0;
+const initialContractAttempts = 0;
+
+export interface CommunityContractData {
+	votes: [Record<string, unknown>];
+	settings: [string, unknown][];
+	balances: { [tokenHolderAddress: string]: number };
+	vault: { [tokenHolderAddress: string]: { balance: number; start: number; end: number }[] };
+}
+
+/**
+ * Oracle class responsible for retrieving the correct data fields from
+ * the ArDrive Community Contract. This class can utilize several different
+ * contract readers and will fallback to other readers if one fails
+ *
+ * @remarks Will begin fetching data from default contract reader on construction
+ */
+export class ArDriveContractOracle implements ContractOracle {
+	constructor(private readonly arweave: Arweave) {
+		// Get contract data upon construction
+		this.getCommunityContract();
+	}
+
+	/**
+	 * Array of contract readers to use as fall back if one fails
+	 * Uses contract reader at index 0 first then descends down the list
+	 */
+	private contractReaders: ContractReader[] = [
+		new VertoContractReader(),
+		new SmartweaveContractReader(this.arweave)
+		// new RedstoneContractReader(this.arweave)
+	];
+	private currentContractReader = initialContractReader;
+	private readContractAttempts = initialContractAttempts;
+
+	private communityContract?: CommunityContractData;
+	private contractPromise?: Promise<CommunityContractData>;
+
+	/** Sets the current contract reader to the next reader in descending order  */
+	fallbackToNextContractReader(): void {
+		const nextContractReaderIndex = this.currentContractReader + 1;
+
+		this.readContractAttempts = initialContractAttempts;
+		this.currentContractReader = nextContractReaderIndex;
+
+		console.log('Falling back to next contract reader..');
+	}
+
+	/**
+	 * Reads a smart contract with the current contract reader
+	 *
+	 * @remarks Will fallback to other contract readers when one fails
+	 */
+	async readContract(txId: string): Promise<unknown> {
+		let contract: unknown;
+
+		try {
+			// Get contract with current contract reader's readContract implementation
+			contract = await this.contractReaders[this.currentContractReader].readContract(txId);
+			return contract;
+		} catch (error) {
+			console.error(`Contract could not fetched: ${error}`);
+
+			this.readContractAttempts++;
+
+			if (this.readContractAttempts >= maxReadContractAttempts) {
+				// Max attempts for contract reader has been reached
+				if (this.currentContractReader === this.contractReaders.length - 1) {
+					// Current contract reader is the last fallback,  throw an error
+					throw new Error(
+						`Max contract read attempts has been reached on the last fallback contract reader..`
+					);
+				}
+
+				// Else fallback to next reader
+				this.fallbackToNextContractReader();
+			} else {
+				console.log('Retrying with current contract reader..');
+			}
+
+			// Retry read contract if no error was thrown
+			return this.readContract(txId);
```

Comment:
Recursive approach is risky and invites anyone who makes changes here to create an infinite loop and blow up the stack. Consider an iterative implementation.